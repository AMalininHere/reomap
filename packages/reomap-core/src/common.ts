import { tile2lat, tile2lng, lng2tile, lat2tile } from './utils/geo-fns';

export class LatLng {
  constructor(
    public readonly lat: number,
    public readonly lng: number,
  ) { }

  equals(other: LatLng) {
    return (
      this.lat === other.lat &&
      this.lng === other.lng
    );
  }
}

export class Point {
  constructor(
    public readonly x: number,
    public readonly y: number,
  ) { }
}

export const TILE_SIZE = 256;

export const absMinLatLng = new LatLng(
  tile2lat(Math.pow(2, 10), 10),
  tile2lng(0, 10)
);

export const absMaxLatLng = new LatLng(
  tile2lat(0, 10),
  tile2lng(Math.pow(2, 10), 10)
);

export function pixelToLatLng(width: number, height: number, zoom: number, center: LatLng, source: Point) {
  const pointDiffX = (source.x - width / 2) / TILE_SIZE;
  const pointDiffY = (source.y - height / 2) / TILE_SIZE;

  const tileX = lng2tile(center.lng, zoom) + pointDiffX;
  const tileY = lat2tile(center.lat, zoom) + pointDiffY;

  return new LatLng(
    Math.max(absMinLatLng.lat, Math.min(absMaxLatLng.lat, tile2lat(tileY, zoom))),
    Math.max(absMinLatLng.lng, Math.min(absMaxLatLng.lng, tile2lng(tileX, zoom)))
  );
}

export function latLngToPixel(width: number, height: number, zoom: number, center: LatLng, source: LatLng) {
  const tileCenterX = lng2tile(center.lng, zoom);
  const tileCenterY = lat2tile(center.lat, zoom);

  const tileX = lng2tile(source.lng, zoom);
  const tileY = lat2tile(source.lat, zoom);

  return new Point(
    (tileX - tileCenterX) * TILE_SIZE + width / 2,
    (tileY - tileCenterY) * TILE_SIZE + height / 2
  );
}
