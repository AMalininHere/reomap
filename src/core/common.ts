import { tile2lat, tile2lng, lng2tile, lat2tile } from './utils/geo-fns';
import { LatLng, Point } from './models';

export const absMinLatLng = new LatLng(
  tile2lat(Math.pow(2, 10), 10),
  tile2lng(0, 10)
);

export const absMaxLatLng = new LatLng(
  tile2lat(0, 10),
  tile2lng(Math.pow(2, 10), 10)
);

export function pixelToLatLng(width: number, height: number, zoom: number, center: LatLng, source: Point) {
  const pointDiffX = (source.x - width / 2) / 256.0;
  const pointDiffY = (source.y - height / 2) / 256.0;

  const tileX = lng2tile(center.lng, zoom) + pointDiffX;
  const tileY = lat2tile(center.lat, zoom) + pointDiffY;

  return new LatLng(
    Math.max(absMinLatLng.lat, Math.min(absMaxLatLng.lat, tile2lat(tileY, zoom))),
    Math.max(absMinLatLng.lng, Math.min(absMaxLatLng.lng, tile2lng(tileX, zoom)))
  );
}

export function latLngToPixel(width: number, height: number, zoom: number, center: LatLng, source: LatLng) {
  const tileCenterX = lng2tile(center.lng, zoom);
  const tileCenterY = lat2tile(center.lat, zoom);

  const tileX = lng2tile(source.lng, zoom);
  const tileY = lat2tile(source.lat, zoom);

  return new Point(
    (tileX - tileCenterX) * 256.0 + width / 2,
    (tileY - tileCenterY) * 256.0 + height / 2
  );
};
