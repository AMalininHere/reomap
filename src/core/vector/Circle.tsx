import React from 'react';
import { LatLng, Point } from '../models';

export interface Props {
  center: LatLng;
  radius: number;
  latLngToPixel: (ll: LatLng) => Point;
}

function Circle(props: Props) {
  const { center, radius, latLngToPixel } = props;
  const point = latLngToPixel(center);

  return (
    <circle fill="#555555" cx={point.x} cy={point.y} r={radius} />
  );
}

export default React.memo(Circle);