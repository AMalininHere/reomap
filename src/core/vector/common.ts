import * as G from 'geojson';
import { LatLng, Point } from '../models';

export interface ElementProps<T extends G.Geometry> {
  geoElement: T;
  latLngToPixel: (ll: LatLng) => Point;
}