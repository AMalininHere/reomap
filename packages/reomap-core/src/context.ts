import React, { useContext } from 'react';
import { LatLng, Point, latLngToPixel, pixelToLatLng } from './common';

export class ContextState {
  constructor(
    public readonly center: LatLng,
    public readonly zoom: number,
    public readonly width: number,
    public readonly height: number,
  ) { }

  public readonly latLngToPixel = (source: LatLng) =>
    latLngToPixel(this.width, this.height, this.zoom, this.center, source);

  public readonly pixelToLatLng = (source: Point) =>
    pixelToLatLng(this.width, this.height, this.zoom, this.center, source);
}

const ctx = React.createContext<ContextState>(null!);

export const MapProvider = ctx.Provider;

export function useMapContext() {
  const data = useContext(ctx);

  return data;
}
