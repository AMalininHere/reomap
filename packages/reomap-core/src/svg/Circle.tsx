import React, { forwardRef, memo } from 'react';
import { useMapContext } from '../context';
import { LatLng } from '../common';

type GeoProps = {
  center: LatLng;
  radius: number;
};

type CirlceProps = Omit<React.SVGProps<SVGCircleElement>, 'cx' | 'cy' | 'r' | 'ref'>;

type Props = CirlceProps & GeoProps;

const Circle = memo(forwardRef<SVGCircleElement, Props>(function Circle(props, ref) {
  const {
    center,
    radius,
    ...circleProps
  } = props;
  const mapContext = useMapContext();

  const point = mapContext.latLngToPixel(center);

  const circlePropsWidthDefaults: CirlceProps = {
    fill: '#555555',
    ...circleProps
  }

  return (
    <circle ref={ref} cx={Math.round(point.x)} cy={Math.round(point.y)} r={radius} {...circlePropsWidthDefaults} />
  );
}));

export default Circle;
