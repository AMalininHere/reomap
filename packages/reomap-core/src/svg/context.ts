import { createContext, useContext } from 'react';
import { ContextState } from '../context';

const ctx = createContext<ContextState>(null!);
ctx.displayName = 'SvgLayerContext';

export const SvgLayerProvider = ctx.Provider;

export function useSvgLayerContext() {
  const data = useContext(ctx);

  return data;
}
