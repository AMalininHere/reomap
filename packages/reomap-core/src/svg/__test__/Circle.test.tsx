import React from 'react';
import { render } from '@testing-library/react';
import Circle from '../Circle';
import { SvgLayerProvider } from '../context';
import { createContextState } from '../../context';
import { latLng } from '../../common';

describe('Circle', () => {
  it('should render', () => {
    const contextState = createContextState(latLng(55.417, 85.276), 6, 800, 600);

    const { container } = render(
      <svg>
        <SvgLayerProvider value={contextState}>
          <Circle center={latLng(53.72271667491848, 87.154541015625)} radius={5} />
        </SvgLayerProvider>
      </svg>
    );

    expect(container).toMatchSnapshot();
  });
});
