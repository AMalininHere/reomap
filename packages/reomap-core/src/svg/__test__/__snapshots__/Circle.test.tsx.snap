// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle should render 1`] = `
<div>
  <svg>
    <circle
      cx="485"
      cy="433"
      fill="#555555"
      r="5"
    />
  </svg>
</div>
`;
